### ISSUE TITLE
Anomaly when trying to detect decreasing fixpoint argument in 8.20

### ISSUE THREAD

**Issue by Janno** [coq/coq#19661]
### Description of the problem

The reproducing example fails with `Cannot guess decreasing argument of fix.` on 8.19 but produces an anomaly on 8.20:
`Error: Anomaly "Uncaught exception Invalid_argument("index out of bounds")."`

<details><summary>Backtrace</summary>

```OCaml
Please report at http://coq.inria.fr/bugs/.
Raised by primitive operation at Inductive.apply_branch in file "fmdeps/coq/kernel/inductive.ml", line 479, characters 11-21
Called from Inductive.check_one_fix.check_rec_call_state in file "fmdeps/coq/kernel/inductive.ml", line 1461, characters 14-28
Called from Inductive.check_one_fix.check_rec_call_stack in file "fmdeps/coq/kernel/inductive.ml" (inlined), line 1222, characters 31-1023
Called from Inductive.check_one_fix.check_rec_call_stack in file "fmdeps/coq/kernel/inductive.ml" (inlined), line 1222, characters 31-1023
Called from Inductive.check_one_fix.check_rec_call in file "fmdeps/coq/kernel/inductive.ml", line 1477, characters 19-70
Called from Inductive.check_one_fix.check_rec_call_stack.(fun) in file "fmdeps/coq/kernel/inductive.ml", line 1228, characters 40-64
Called from Inductive.check_one_fix.check_rec_call_stack.(fun) in file "fmdeps/coq/kernel/inductive.ml" (inlined), line 1227, characters 33-231
Called from Stdlib__Array.fold_right in file "array.ml" (inlined), line 184, characters 9-30
Called from Inductive.check_one_fix.check_rec_call_stack in file "fmdeps/coq/kernel/inductive.ml" (inlined), line 1227, characters 16-247
Called from Inductive.check_one_fix.check_rec_call_stack in file "fmdeps/coq/kernel/inductive.ml", line 1227, characters 16-247
Called from Inductive.check_one_fix.check_rec_call_stack in file "fmdeps/coq/kernel/inductive.ml" (inlined), line 1222, characters 31-1023
Called from Inductive.check_one_fix.check_rec_call_stack in file "fmdeps/coq/kernel/inductive.ml" (inlined), line 1222, characters 31-1023
Called from Inductive.check_one_fix.check_rec_call in file "fmdeps/coq/kernel/inductive.ml", line 1477, characters 19-70
Called from Inductive.check_one_fix in file "fmdeps/coq/kernel/inductive.ml", line 1480, characters 24-50
Called from Pretyping.search_guard.(fun) in file "fmdeps/coq/pretyping/pretyping.ml", line 128, characters 23-47
Called from Stdlib__List.iter in file "list.ml" (inlined), line 110, characters 12-15
Called from Stdlib__List.iter in file "list.ml" (inlined), line 108, characters 13-64
Called from Pretyping.search_guard in file "fmdeps/coq/pretyping/pretyping.ml", line 117, characters 16-686
Called from Declare.make_recursive_body in file "fmdeps/coq/vernac/declare.ml", line 736, characters 16-73
Called from Declare.make_mutual_bodies in file "fmdeps/coq/vernac/declare.ml", line 756, characters 22-76
Called from ComFixpoint.declare_recursive in file "fmdeps/coq/vernac/comFixpoint.ml", line 278, characters 6-131
Called from Vernacentries.vernac_fixpoint in file "fmdeps/coq/vernac/vernacentries.ml", line 1104, characters 16-98
Called from Vernacentries.translate_pure_vernac.(fun) in file "fmdeps/coq/vernac/vernacentries.ml" (inlined), line 2467, characters 61-92
Called from Vernacentries.translate_pure_vernac.(fun) in file "fmdeps/coq/vernac/vernacentries.ml" (inlined), line 2467, characters 24-94
Called from Vernactypes.vtmodifyprogram.(fun) in file "fmdeps/coq/vernac/vernactypes.ml" (inlined), line 197, characters 30-40
Called from Vernactypes.typed_vernac.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 170, characters 65-71
Called from Vernactypes.run.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 164, characters 15-32
Called from Vernactypes.combine_runners.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 126, characters 16-23
Called from Vernactypes.OpaqueAccess.runner.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 114, characters 30-34
Called from Vernactypes.combine_runners.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 125, characters 14-100
Called from Vernactypes.combine_runners.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 126, characters 16-23
Called from Vernactypes.Proof.runner.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 69, characters 30-34
Called from Vernactypes.combine_runners.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 125, characters 14-100
Called from Vernactypes.Prog.runner.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 30, characters 20-24
Called from Vernactypes.combine_runners.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 124, characters 12-173
Called from Vernactypes.combine_runners.(fun) in file "fmdeps/coq/vernac/vernactypes.ml", line 124, characters 12-173
Called from Vernactypes.run in file "fmdeps/coq/vernac/vernactypes.ml", line 163, characters 14-96
Called from Vernacinterp.interp_expr in file "fmdeps/coq/vernac/vernacinterp.ml", line 120, characters 60-157
Called from Vernacinterp.interp_control.(fun) in file "fmdeps/coq/vernac/vernacinterp.ml" (inlined), line 154, characters 11-46
Called from Synterp.with_generic_atts in file "fmdeps/coq/vernac/synterp.ml" (inlined), line 71, characters 12-19
Called from Vernacinterp.interp_control.(fun) in file "fmdeps/coq/vernac/vernacinterp.ml" (inlined), line 153, characters 24-126
Called from Vernacinterp.interp_control.(fun) in file "fmdeps/coq/vernac/vernacinterp.ml", line 153, characters 24-126
Called from Flags.with_modified_ref in file "fmdeps/coq/lib/flags.ml" (inlined), line 17, characters 14-17
Called from Flags.with_option in file "fmdeps/coq/lib/flags.ml" (inlined), line 26, characters 24-78
Called from Flags.silently in file "fmdeps/coq/lib/flags.ml", line 65, characters 19-40
Re-raised at Exninfo.iraise in file "fmdeps/coq/clib/exninfo.ml", line 81, characters 4-38
Called from Vernacinterp.interp_gen in file "fmdeps/coq/vernac/vernacinterp.ml" (inlined), line 197, characters 16-41
Called from Vernacinterp.interp.(fun) in file "fmdeps/coq/vernac/vernacinterp.ml", line 211, characters 54-111
Re-raised at Exninfo.iraise in file "fmdeps/coq/clib/exninfo.ml", line 81, characters 4-38
Called from NewProfile.profile in file "fmdeps/coq/lib/newProfile.ml" (inlined), line 201, characters 32-36
Called from Vernacinterp.interp in file "fmdeps/coq/vernac/vernacinterp.ml" (inlined), line 211, characters 15-115
Called from Vernacinterp.interp in file "fmdeps/coq/vernac/vernacinterp.ml", line 211, characters 15-115
Called from Stm.Reach.known_state.reach.(fun) in file "fmdeps/coq/stm/stm.ml", line 2029, characters 16-43
Called from Stm.State.define in file "fmdeps/coq/stm/stm.ml", line 959, characters 6-10
Re-raised at Exninfo.iraise in file "fmdeps/coq/clib/exninfo.ml", line 81, characters 4-38
Called from Stm.Reach.known_state.reach in file "fmdeps/coq/stm/stm.ml", line 2176, characters 4-105
Called from Stm.observe in file "fmdeps/coq/stm/stm.ml", line 2268, characters 4-60
Re-raised at Exninfo.iraise in file "fmdeps/coq/clib/exninfo.ml", line 81, characters 4-38
Called from Vernac.interp_vernac in file "fmdeps/coq/toplevel/vernac.ml", line 83, characters 29-50
Re-raised at Exninfo.iraise in file "fmdeps/coq/clib/exninfo.ml", line 81, characters 4-38
Called from Vernac.load_vernac_core.loop.(fun) in file "fmdeps/coq/toplevel/vernac.ml" (inlined), line 138, characters 28-57
Called from Flags.with_modified_ref in file "fmdeps/coq/lib/flags.ml" (inlined), line 17, characters 14-17
Called from Flags.with_option in file "fmdeps/coq/lib/flags.ml" (inlined), line 26, characters 24-78
Called from Flags.silently in file "fmdeps/coq/lib/flags.ml" (inlined), line 65, characters 19-40
Called from Vernac.load_vernac_core.loop.(fun) in file "fmdeps/coq/toplevel/vernac.ml" (inlined), line 138, characters 13-61
Called from Vernac.load_vernac_core.loop.(fun) in file "fmdeps/coq/toplevel/vernac.ml", line 138, characters 13-61
Re-raised at Exninfo.iraise in file "fmdeps/coq/clib/exninfo.ml", line 81, characters 4-38
Called from Util.try_finally in file "fmdeps/coq/lib/util.ml" (inlined), line 133, characters 16-19
Called from Vernac.load_vernac_core.loop in file "fmdeps/coq/toplevel/vernac.ml" (inlined), line 127, characters 8-774
Called from Vernac.load_vernac_core.loop in file "fmdeps/coq/toplevel/vernac.ml", line 127, characters 8-774
Re-raised at Exninfo.iraise in file "fmdeps/coq/clib/exninfo.ml", line 81, characters 4-38
Called from Util.try_finally in file "fmdeps/coq/lib/util.ml" (inlined), line 137, characters 6-32
Called from Vernac.load_vernac_core.loop in file "fmdeps/coq/toplevel/vernac.ml", line 127, characters 8-774
Called from Vernac.load_vernac_core.loop in file "fmdeps/coq/toplevel/vernac.ml" (inlined), line 109, characters 15-1023
Called from Vernac.load_vernac_core in file "fmdeps/coq/toplevel/vernac.ml", line 150, characters 6-19
Re-raised at Exninfo.iraise in file "fmdeps/coq/clib/exninfo.ml", line 81, characters 4-38
Called from Vernac.load_vernac_core in file "fmdeps/coq/toplevel/vernac.ml", line 154, characters 4-28
Called from Ccompile.compile in file "fmdeps/coq/toplevel/ccompile.ml", line 68, characters 18-78
Called from Coqc.coqc_main in file "fmdeps/coq/toplevel/coqc.ml", line 38, characters 2-100
```
</details>

### Small Coq file to reproduce the bug

```coq
From Coq Require Import Utf8.

Inductive tele : Type :=
  | TeleO : tele
  | TeleS {X} (binder : X -> tele) : tele.

Fixpoint tele_fun (TT : tele) (T : Type) : Type :=
  match TT with
  | TeleO => T
  | TeleS b => forall x, tele_fun (b x) T
  end.

Notation "TT -t> A" := (tele_fun TT A) (at level 99, A at level 200, right associativity).

Record tele_arg_cons {X : Type} (f : X → Type) : Type := TeleArgCons
  { tele_arg_head : X;
    tele_arg_tail : f tele_arg_head }.
Global Arguments TeleArgCons {_ _} _ _.

Fixpoint tele_arg (t : tele) : Type :=
  match t with
  | TeleO => unit
  | TeleS f => tele_arg_cons (λ x, tele_arg (f x))
  end.

Fixpoint tele_app {TT : tele} {U} : (TT -t> U) → tele_arg TT → U :=
  match TT as TT return (TT -t> U) → tele_arg TT → U with
  | TeleO => λ F _, F
  | TeleS r => λ (F : TeleS r -t> U) '(TeleArgCons x b),
      tele_app (F x) b
  end.
Global Arguments tele_app {!_ _} & _ !_ /.

Axiom PROP: Type.
Axiom T : PROP.
Axiom atomic_commit :
  forall {TB : tele}
         (_ : forall (_ : tele_arg TB), PROP)
         (_ : forall (_ : tele_arg TB), PROP),
    PROP.
Set Debug "backtrace".
Fixpoint do_after_shatter  (next : nat) (d : nat) : PROP :=
  (@atomic_commit
      (@TeleS _ (fun l' => TeleO))
      ((@tele_app (@TeleS _ (fun l' => TeleO))
          _ (fun l' => T)))
      ((@tele_app (@TeleS _ (fun l' =>  TeleO))
          _ (fun l' =>
                      do_after_shatter (l'-1) d
                    )))).
```

### Version of Coq where this bug occurs

8.20

### Last version of Coq where the bug did not occur

8.19

**Comment by Janno** [comment]
git bisect points to
```
140908d6a27559b141203500da0681fd5ebb23e3 is the first bad commit
commit 140908d6a27559b141203500da0681fd5ebb23e3
Author: Hugo Herbelin <Hugo.Herbelin@inria.fr>
Date:   Sun Aug 27 23:31:12 2023 +0200

    Grant wish #16040: inline uniform arguments of an inner fix in guard checking.
    
    We proceed in three steps on the expansion of the relevant body of a block of fix:
    - first find the arguments of the recursive calls that are uniform
    - then drop those arguments
    - in addition to the main argument of the inner fix, we propagate the
      recursive specification of the arguments that are uniform towards
      the body of the fixpoints.
```

/cc @herbelin

**Comment by herbelin** [comment]
That's an interestingly and elaborate example. The wrong stack of arguments is passed to `tele_app` of `TeleS` but more debugging would be needed to fully understand the problem.

**Comment by herbelin** [comment]
Fixed in #19671. Thanks for reporting (it has still to be checked whether it could be critical).
